File: macos/Scanner/Sources/Scanner/Rules/AESMCP007.swift
public struct AESMCP007: Rule {
    public let id = "AES-MCP-007"
    public let severity: Severity = .medium

    public init() {}

    public func evaluate(_ facts: ScanFacts) -> [Finding] {
        facts.mcpServers.compactMap { server in
            guard !server.disabled else {
                return nil
            }
            guard PackageVersion.isPackageRunner(server.commandName) else {
                return nil
            }
            guard !PackageVersion.hasPinnedVersion(args: server.args) else {
                return nil
            }
            return finding(app: server.appId, serverName: server.name, affectedPath: server.configPath)
        }
    }
}

File: windows/Scanner/Rules/AESMCP007.cs
using AIExposureScanner.Scanner.Models;

namespace AIExposureScanner.Scanner.Rules;

public sealed class AESMCP007 : IRule
{
    public string Id => "AES-MCP-007";
    public Severity Severity => Severity.Medium;

    public IReadOnlyList<Finding> Evaluate(ScanFacts facts) =>
        facts.McpServers
            .Where(server => !server.Disabled)
            .Where(server => PackageVersion.IsPackageRunner(server.CommandName()))
            .Where(server => !PackageVersion.HasPinnedVersion(server.Args))
            .Select(server => this.Finding(server.AppId, server.Name, affectedPath: server.ConfigPath))
            .ToArray();
}
